src/sidebar/store/use-store.js

```diff
@@ -97,3 +97,135 @@ export default function useStore(callback) {
+  matches(method, args) {
+    return (
+      this.method === method && this.args.every((value, i) => args[i] === value)
+    );
+  }
+}
+
+/**
+ * Return a wrapper around the `store` service that UI components can use to
+ * extract data from the store and call actions on it.
+ *
+ * Unlike using the `store` service directly, the wrapper tracks what data from
+ * the store the current component uses and re-renders the component when it
+ * changes.
+ *
+ * The returned wrapper has the same API as the store itself.
+ *
+ * @example
+ *   function MyComponent() {
+ *     const store = useStoreProxy();
+ *     const currentUser = store.currentUser();
+ *
+ *     return (
+ *       <div>
+ *         Current user: {currentUser}.
+ *         <button onClick={() => store.logOut()}>Log out</button>
+ *       </div>
+ *     );
+ *   }
+ *
+ * @return {SidebarStore}
+ */
+export function useStoreProxy() {
+  const store = useService('store');
+
+  // Hack to trigger a component re-render.
+  const [, forceUpdate] = useReducer(x => x + 1, 0);
+
+  // Cache of store method calls made by the current UI component and associated
+  // results. This is currently just an array on the assumption that it will
+  // only have a small number of entries. It could be changed to a map keyed
+  // by method name to handle many entries better.
+  const cacheRef = useRef(/** @type {CacheEntry[]} */ ([]));
+  const cache = cacheRef.current;
+
+  // Create the wrapper around the store.
+  const proxy = useRef(/** @type {SidebarStore|null} */ (null));
+  if (!proxy.current) {
+    // Cached method wrappers.
+    const wrappedMethods = {};
+
+    /**
+     * @param {typeof store} target
+     * @param {string} prop
+     */
+    const get = (target, prop) => {
+      const original = target[prop];
+      if (typeof original !== 'function') {
+        return original;
+      }
+
+      // Check for pre-existing method wrapper.
+      let wrapped = wrappedMethods[prop];
+      if (wrapped) {
+        return wrapped;
+      }
+
+      // Create method wrapper.
+      wrapped = (...args) => {
+        const cacheEntry = cache.find(entry => entry.matches(prop, args));
+        if (cacheEntry) {
+          return cacheEntry.result;
+        }
+
+        // Call the original method. It may be a selector that does not modify
+        // the store but returns a result, or an action that modifies the state
+        // but does not return a result.
+        const result = original.call(target, ...args);
+        if (result !== undefined) {
+          cache.push(new CacheEntry(prop, args, result));
+        }
+        return result;
+      };
+      wrappedMethods[prop] = wrapped;
+
+      return wrapped;
+    };
+
+    proxy.current = new Proxy(store, { get });
+  }
+
+  // Register a subscriber which clears cache and re-renders component when
+  // relevant store state changes.
+  useEffect(() => {
+    const cleanup = store.subscribe(() => {
+      const isCacheInvalid = cache.some(
```

Comment:
Gimme a hand here (running low on time and don't want to unravel `useStore` internals right now)...does this cache-clearing of all "registered" store selectors for the calling component ultimately result in selectors getting recomputed more often than the previous implementation? The same? Apples and oranges?

Reply:
It _may_ be slightly more efficient in the case where a re-render needs to happen. Previously we would always re-evaluate every selector used by a component when anything in the store changed (ie. we would re-run all the `useStore` hook callbacks). We will now evaluate the selector calls in the order they occurred and bail out early if we detect that any are invalid.